DataReader read and take APIs [10283] (#222)
eProsima/Fast-RTPS-docs: docs/fastdds/dds_layer/subscriber/dataReader/readingData.rst

The application can access and consume the data values received on the :ref:`dds_layer_subscriber_dataReader`
by *reading* or *taking*.

 * **Reading** is done with any of the following member functions:

   * |DataReader::read_next_sample-api| reads the next, non-previously accessed data value available
     on the DataReader, and stores it in the provided data buffer.

>    * |DataReader::read-api|, |DataReader::read_instance-api| and |DataReader::read_next_instance-api|
>      provide mechanisms to get a collection of samples matching certain conditions.

Reviewer: ```suggestion
   * |DataReader::read-api|, |DataReader::read_instance-api|, and |DataReader::read_next_instance-api|
     provide mechanisms to get a collection of samples matching certain conditions.
```

 * **Taking** is done with any of the following member functions:

>    * |DataReader::take_next_sample-api| reads the next, non-previously accessed data value available on the DataReader,
>      and stores it in the provided data buffer.

Reviewer: ```suggestion
   * |DataReader::take_next_sample-api| reads the next data value available on the DataReader,
     previously accessed or not, and stores it in the provided data buffer.
```

Author: Dismissed after offline conversation

>    * |DataReader::take-api|, |DataReader::take_instance-api| and |DataReader::take_next_instance-api|
>      provide mechanisms to get a collection of samples matching certain conditions.

Reviewer: ```suggestion
   * |DataReader::take-api|, |DataReader::take_instance-api|, and |DataReader::take_next_instance-api|
     provide mechanisms to get a collection of samples matching certain conditions.
```

When there is no data in the DataReader matching the required conditions, all the operations will return
``NO_DATA`` and output parameter will remain unchanged.

In addition to the data values, the data access operations also provide SampleInfo instances with additional
information that help interpreting the returned data values, like the originating
:ref:`dds_layer_publisher_dataWriter` or the publication time stamp.
Please, refer to the :ref:`dds_layer_subscriber_sampleInfo` section for an extensive description of its contents.

.. _dds_layer_subscriber_accessreceived_loans:

Loaning and Returning Data and SampleInfo Sequences
---------------------------------------------------

The |DataReader::read-api| and |DataReader::take-api| operations (and their variants) return information to the
application in two sequences:

 * Received DDS data samples in a sequence of the data type
 * Corresponding information about each DDS sample in a SampleInfo sequence

These sequences are parameters that are passed by the application code into the
|DataReader::read-api| and |DataReader::take-api| operations.

> When empty sequences (sequences that are initialized but have a maximum length of 0) are used, the middleware will

Reviewer: ```suggestion
When the passed sequences are empty (they are initialized but have a maximum length of 0), the middleware will
```

I find it easier to read it this way

fill those sequences with memory directly loaned from the receive queue itself.
There is no copying of the data or SampleInfo when the contents of the sequences are loaned.
This is certainly the most efficient way for the application code to retrieve the data.

> When doing so, however, the code must return the loaned sequences back to the middleware so that they can be reused
> by the receive queue.

Reviewer: ```suggestion
When doing so, however, the code must return the loaned sequences back to the middleware, so that they can be reused
by the receive queue.
```

If the application does not return the loan by calling the |DataReader::return_loan-api| operation, then Fast DDS
will eventually run out of memory to store DDS data samples received from the network for that DataReader.
See the code below for an example of borrowing and returning loaned sequences.

.. literalinclude:: /../code/DDSCodeTester.cpp
   :language: c++
   :start-after: //DDS_DATAREADER_LOAN_SEQUENCES
   :end-before: //!
   :dedent: 8

.. _dds_layer_subscriber_accessreceived_data:

Processing returned data
------------------------

After calling the |DataReader::read-api| or |DataReader::take-api| operations, accessing the data on the returned
sequences is quite easy.

> The sequences API provide a **length()** operation returning the number of elements in the collections.

Reviewer: ```suggestion
The sequences API provides a **length()** operation returning the number of elements in the collections.
```
